nixos/cosmic: add missing dependencies

kajusnau opened the pull request:
#### Added missing dependencies for some cosmic tools

-  cosmic-settings network page
   - **_networkmanagerapplet_** for `nm-connection-editor`
   https://github.com/pop-os/cosmic-settings/blob/ae5f151932f936f3dd11a8467b8733d0fd260cdb/cosmic-settings/src/pages/networking/mod.rs#L375
- cosmic-settings-daemon
   - **_pulseaudio_** for `pactl`
   https://github.com/pop-os/cosmic-settings-daemon/blob/8616c40d235164779cd3f2ceec1fe9b2b4aceb40/src/pulse.rs#L16
   - **_glib_** for `gsettings`
   https://github.com/pop-os/cosmic-settings-daemon/blob/8616c40d235164779cd3f2ceec1fe9b2b4aceb40/src/theme.rs#L494

#### Patched **networkmanagerapplet**

- Added a small patch to disable autostart in COSMIC, following the [behavior of Pop_OS](https://github.com/pop-os/network-manager-applet/) (thanks @michaelBelsanti, @Pandapip1)

<!--
^ Please summarise the changes you have done and explain why they are necessary here ^

For package updates please link to a changelog or describe changes, this helps your fellow maintainers discover breaking updates.
For new packages please briefly describe the package or provide a link to its homepage.
-->

## Things done

<!-- Please check what applies. Note that these are not hard requirements but merely serve as information for reviewers. -->

- Built on platform:
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- Tested, as applicable:
  - [x] [NixOS tests] in [nixos/tests].
  - [ ] [Package tests] at `passthru.tests`.
  - [ ] Tests in [lib/tests] or [pkgs/test] for functions and "core" functionality.
- [x] Ran `nixpkgs-review` on this PR. See [nixpkgs-review usage].
- [ ] Tested basic functionality of all binary files, usually in `./result/bin/`.
- Nixpkgs Release Notes
  - [ ] Package update: when the change is major or breaking.
- NixOS Release Notes
  - [ ] Module addition: when adding a new NixOS module.
  - [ ] Module update: when the change is significant.
- [x] Fits [CONTRIBUTING.md], [pkgs/README.md], [maintainers/README.md] and other READMEs.

[NixOS tests]: https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests
[Package tests]: https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#package-tests
[nixpkgs-review usage]: https://github.com/Mic92/nixpkgs-review#usage

[CONTRIBUTING.md]: https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md
[lib/tests]: https://github.com/NixOS/nixpkgs/blob/master/lib/tests
[maintainers/README.md]: https://github.com/NixOS/nixpkgs/blob/master/maintainers/README.md
[nixos/tests]: https://github.com/NixOS/nixpkgs/blob/master/nixos/tests
[pkgs/README.md]: https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md
[pkgs/test]: https://github.com/NixOS/nixpkgs/blob/master/pkgs/test

---

Add a :+1: [reaction] to [pull requests you find important].

[reaction]: https://github.blog/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/
[pull requests you find important]: https://github.com/NixOS/nixpkgs/pulls?q=is%3Aopen+sort%3Areactions-%2B1-desc

kajusnau commented:
**networkmanagerapplet** package installs its own autostart entry at `etc/xdg/autostart/nm-applet.desktop`, which will then be linked if [xdg.autostart.enable](https://search.nixos.org/options?channel=unstable&show=xdg.autostart.enable&query=xdg.autostart) is set to true (the default)

This is slightly problematic - COSMIC already provides its own network applet by default, meaning two network applets will be running at the same time.

We could address this by explicitly disabling the **nm-applet** autostart service via
```
"app-nm\\x2dapplet@autostart" = {
  enable = false;
};
```
but this approach is a little hacky

Another alternative is to patch **networkmanagerapplet** by adding `NotShowIn=COSMIC` to the `nm-applet.desktop`, but I'm not fully certain how this works.

I'm open to ideas and discussion here 🙂

kajusnau commented on nixos/modules/services/desktop-managers/cosmic.nix:
```diff
@@ -94,11 +94,21 @@ in
+            postPatch = (oldAttrs.postPatch or "") + ''
+              sed -i '/^NotShowIn=/ s/;*$/;COSMIC;/' nm-applet.desktop.in
+            '';
+          }))
```
Thanks for the info, @Pandapip1 

I agree that the applet can be patched directly within the scope of this PR, as a separate commit.
I’ll update the PR with these changes over the weekend 🫡

nixos-discourse commented:
This pull request has been mentioned on **NixOS Discourse**. There might be relevant details there:

https://discourse.nixos.org/t/prs-already-reviewed/2617/2592

HeitorAugustoLN commented:
## `nixpkgs-review` result

Generated using [`nixpkgs-review`](https://github.com/Mic92/nixpkgs-review).

Command: `nixpkgs-review pr 449426`
Commit: `464a3365026fb8bcab243d2a6756ce4f9761730a`

---
### `x86_64-linux`
<details>
  <summary>:fast_forward: 2 packages blacklisted:</summary>
  <ul>
    <li>nixos-install-tools</li>
    <li>tests.nixos-functions.nixos-test</li>
  </ul>
</details>
<details>
  <summary>:white_check_mark: 14 packages built:</summary>
  <ul>
    <li>budgie-control-center</li>
    <li>budgie-control-center.debug</li>
    <li>firewalld-gui</li>
    <li>gnome-control-center</li>
    <li>gnome-control-center.debug</li>
    <li>networkmanagerapplet</li>
    <li>networkmanagerapplet.man</li>
    <li>pantheon.elementary-greeter</li>
    <li>pantheon.switchboard-plug-network</li>
    <li>pantheon.switchboard-with-plugs</li>
    <li>pantheon.wingpanel-applications-menu</li>
    <li>pantheon.wingpanel-with-indicators</li>
    <li>phosh</li>
    <li>phosh-mobile-settings</li>
  </ul>
</details>

---
### `aarch64-linux`
<details>
  <summary>:fast_forward: 2 packages blacklisted:</summary>
  <ul>
    <li>nixos-install-tools</li>
    <li>tests.nixos-functions.nixos-test</li>
  </ul>
</details>
<details>
  <summary>:white_check_mark: 14 packages built:</summary>
  <ul>
    <li>budgie-control-center</li>
    <li>budgie-control-center.debug</li>
    <li>firewalld-gui</li>
    <li>gnome-control-center</li>
    <li>gnome-control-center.debug</li>
    <li>networkmanagerapplet</li>
    <li>networkmanagerapplet.man</li>
    <li>pantheon.elementary-greeter</li>
    <li>pantheon.switchboard-plug-network</li>
    <li>pantheon.switchboard-with-plugs</li>
    <li>pantheon.wingpanel-applications-menu</li>
    <li>pantheon.wingpanel-with-indicators</li>
    <li>phosh</li>
    <li>phosh-mobile-settings</li>
  </ul>
</details>

GaetanLepage reviewed (commented):
a

Pandapip1 commented on pkgs/by-name/ne/networkmanagerapplet/package.nix:
```diff
@@ -73,6 +73,9 @@ stdenv.mkDerivation rec {
     patchShebangs meson_post_install.py
+
+    # Prevent applet from autostarting in COSMIC, which has its own built-in network applet
+    sed -i '/^NotShowIn=/ s/;*$/;COSMIC;/' nm-applet.desktop.in
```
I'd lean more towards keeping `sed` here but I can see the argument for both. The `substituteInPlace` version isn't going to offend anyone though.

GaetanLepage replied:
I could be wrong, but wouldn't `substituteInPlace` fail more loudly if the substitution could not be performed in the future?

kajusnau committed:
networkmanagerapplet: disable autostart in COSMIC DE

kajusnau committed:
nixos/cosmic: add missing dependencies

- added missing dependencies for some cosmic tools
  - cosmic-settings network page
    'networkmanagerapplet' for 'nm-connection-editor'
  - cosmic-settings-daemon
    'pulseaudio' for 'pactl'
    'glib' for 'gsettings'

GaetanLepage reviewed (approved):
LGTM!

GaetanLepage commented:
## `nixpkgs-review` result

Generated using [`nixpkgs-review`](https://github.com/Mic92/nixpkgs-review).

Command: `nixpkgs-review pr 449426`
Commit: `fdf261415571131a533382b99995bb13036d37d8`

---
### `x86_64-linux`
<details>
  <summary>:fast_forward: 2 packages blacklisted:</summary>
  <ul>
    <li>nixos-install-tools</li>
    <li>tests.nixos-functions.nixos-test</li>
  </ul>
</details>
<details>
  <summary>:white_check_mark: 14 packages built:</summary>
  <ul>
    <li>budgie-control-center</li>
    <li>budgie-control-center.debug</li>
    <li>firewalld-gui</li>
    <li>gnome-control-center</li>
    <li>gnome-control-center.debug</li>
    <li>networkmanagerapplet</li>
    <li>networkmanagerapplet.man</li>
    <li>pantheon.elementary-greeter</li>
    <li>pantheon.switchboard-plug-network</li>
    <li>pantheon.switchboard-with-plugs</li>
    <li>pantheon.wingpanel-applications-menu</li>
    <li>pantheon.wingpanel-with-indicators</li>
    <li>phosh</li>
    <li>phosh-mobile-settings</li>
  </ul>
</details>

---
### `aarch64-linux`
<details>
  <summary>:fast_forward: 2 packages blacklisted:</summary>
  <ul>
    <li>nixos-install-tools</li>
    <li>tests.nixos-functions.nixos-test</li>
  </ul>
</details>
<details>
  <summary>:white_check_mark: 14 packages built:</summary>
  <ul>
    <li>budgie-control-center</li>
    <li>budgie-control-center.debug</li>
    <li>firewalld-gui</li>
    <li>gnome-control-center</li>
    <li>gnome-control-center.debug</li>
    <li>networkmanagerapplet</li>
    <li>networkmanagerapplet.man</li>
    <li>pantheon.elementary-greeter</li>
    <li>pantheon.switchboard-plug-network</li>
    <li>pantheon.switchboard-with-plugs</li>
    <li>pantheon.wingpanel-applications-menu</li>
    <li>pantheon.wingpanel-with-indicators</li>
    <li>phosh</li>
    <li>phosh-mobile-settings</li>
  </ul>
</details>
